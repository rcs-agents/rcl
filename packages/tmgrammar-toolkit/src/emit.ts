/**
 * Grammar processing and emit functions
 * Handles validation, repository management, and output generation
 */

import { readFile } from "node:fs/promises";
import { loadWASM, OnigRegExp } from "onigasm";
import { dirname, resolve } from "node:path";
import * as plist from "plist";
import type { Grammar, Rule, EmitOptions, MatchRule, BeginEndRule, IncludeRule, BasicIncludePattern } from './types.js';
import { meta } from './types.js';

let initialized = false;
async function initialize() {
  if (!initialized) {
    const onigasmPath = import.meta.resolve("onigasm");
    const wasmPath = resolve(dirname(new URL(onigasmPath).pathname), "onigasm.wasm");
    const wasm = await readFile(wasmPath);
    await loadWASM(wasm.buffer as any);
    initialized = true;
  }
}

/**
 * Emit the given grammar to JSON.
 */
export async function emitJSON(grammar: Grammar, options: EmitOptions = {}): Promise<string> {
  await initialize();
  const indent = 2;
  const processed = await processGrammar(grammar, options);
  return JSON.stringify(processed, undefined, indent);
}

/**
 * Emit the given grammar to PList XML
 */
export async function emitPList(grammar: Grammar, options: EmitOptions = {}): Promise<string> {
  await initialize();
  const processed = await processGrammar(grammar, options);
  return plist.build(processed);
}

/**
 * Convert the grammar from our representation to the tmlanguage schema.
 * Perform validation in the process.
 */
async function processGrammar(grammar: Grammar, options: EmitOptions): Promise<any> {
  await initialize();

  const internalRepositoryMap = new Map<string, [Rule, any]>(); // Maps rule key to [originalRule, processedRuleDefinition]
  const grammarNameLower = grammar.name.toLowerCase();

  // 1. Pre-populate and process all rules from repositoryItems
  if (grammar.repositoryItems) {
    for (const rule of grammar.repositoryItems) {
      if ('key' in rule) { // Only rules with keys can be in the repository
        const keyedRule = rule as MatchRule | BeginEndRule | IncludeRule;
        if (!internalRepositoryMap.has(keyedRule.key)) {
          const entry: [Rule, any] = [keyedRule, undefined]; // Placeholder for cyclic dependencies
          internalRepositoryMap.set(keyedRule.key, entry);
          // Process the rule definition itself and store it in entry[1]
          entry[1] = processNode(keyedRule, options, grammarNameLower, internalRepositoryMap, true);
        } else {
          const existingEntry = internalRepositoryMap.get(keyedRule.key);
          if (existingEntry && existingEntry[0] !== keyedRule) {
            throw new Error(
              `Duplicate key found in repositoryItems: '${keyedRule.key}'. \nThe same key is used for different rule objects.`);
          }
        }
      }
    }
  } else {
    // eslint-disable-next-line no-console
    console.warn(
      'Warning: grammar.repositoryItems not provided. Repository discovery will be on-the-fly, which might be incomplete for complex grammars. ' +
      'It is recommended to list all keyed rules in repositoryItems for reliable repository generation.'
    );
  }

  // 2. Process the main grammar structure. processNode for the grammar object will handle its top-level fields
  // and its 'patterns' array. The 'repositoryItems' field itself will be skipped during this processing.
  const outputGrammarStructure = processNode(grammar, options, grammarNameLower, internalRepositoryMap, false);

  // 3. Construct the final repository object for the output from our populated internalRepositoryMap
  const finalRepositoryObject: any = {};
  const sortedKeys = Array.from(internalRepositoryMap.keys()).sort(); // For consistent output
  for (const key of sortedKeys) {
    const entry = internalRepositoryMap.get(key);
    if (entry && entry[1] !== undefined) { // entry[1] holds the processed rule definition
      finalRepositoryObject[key] = entry[1];
    }
  }
  outputGrammarStructure.repository = finalRepositoryObject;

  return outputGrammarStructure;
}

// Added internalRepo parameter, and isProcessingRepositoryItem flag
function processNode(
  node: any, 
  options: EmitOptions, 
  grammarName: string, 
  currentRepository: Map<string, [Rule, any]>, 
  isRepositoryItemContext: boolean
): any {
  if (typeof node !== 'object' || node === null) {
    return node;
  }
  if (Array.isArray(node)) {
    return node.map((n) => processNode(n, options, grammarName, currentRepository, isRepositoryItemContext));
  }

  const outputNode: any = {};
  for (const originalKey in node) {
    const value = node[originalKey];
    let processedValue = value;

    switch (originalKey) {
      case "key": // Drop 'key' from the output of individual rule definitions
        if (!isRepositoryItemContext) { // Keep key if it's part of a structure not meant for repo (e.g. captures)
            outputNode[originalKey] = value;
        }
        break;
      case "repositoryItems": // Always drop 'repositoryItems' from the main grammar output
        break;
      case "scope": {
        const ruleKeyForMeta = node.key || 'unknown'; 
        outputNode.name = value === meta
          ? `meta.${ruleKeyForMeta}.${grammarName}`
          : String(value);
        break;
      }
      case "begin":
      case "end":
      case "match":
      case "while":
      case "firstLineMatch":
      case "foldingStartMarker":
      case "foldingStopMarker":
        if (value instanceof RegExp) {
          processedValue = value.source;
        }
        validateRegexp(processedValue, node, originalKey, options);
        outputNode[originalKey] = processedValue;
        break;
      case "patterns":
        outputNode[originalKey] = processPatterns(value, options, grammarName, currentRepository);
        break;
      default:
        outputNode[originalKey] = processNode(value, options, grammarName, currentRepository, false); // Children are not repo items context by default
        break;
    }
  }
  return outputNode;
}

function processPatterns(
  rules: Rule[], 
  options: EmitOptions, 
  grammarName: string, 
  currentRepository: Map<string, [Rule, any]>
): any[] {
  const processedPatternsArray: any[] = [];
  for (const rule of rules) {
    if ('include' in rule && !('key' in rule)) { // BasicIncludePattern like { include: "#foo" }
      processedPatternsArray.push(rule); 
      continue;
    }

    const keyedRule = rule as MatchRule | BeginEndRule | IncludeRule;

    if (!keyedRule.key) {
      // eslint-disable-next-line no-console
      console.warn('Processing a rule without a key directly in patterns (will be inlined):科研', rule);
      processedPatternsArray.push(processNode(rule, options, grammarName, currentRepository, false));
      continue;
    }

    // Rule has a key. It should be part of the repository.
    if (!currentRepository.has(keyedRule.key)) {
      // This implies the rule was in a 'patterns' array, had a key, but was NOT in grammar.repositoryItems.
      // Process it now and add to repository (fallback if repositoryItems is not used or is incomplete).
      // eslint-disable-next-line no-console
      console.warn(
        `Rule with key '${keyedRule.key}' was encountered in a 'patterns' array but not pre-processed via 'repositoryItems'.
    Processing it now for the repository.`
      );
      const entry: [Rule, any] = [keyedRule, undefined]; 
      currentRepository.set(keyedRule.key, entry);
      entry[1] = processNode(keyedRule, options, grammarName, currentRepository, true); // Process as if it were a repo item
    } else {
      const existingEntry = currentRepository.get(keyedRule.key);
      if (existingEntry && existingEntry[0] !== keyedRule) {
          // Check if they are structurally equivalent (same content, different object instances)
          const existingRule = existingEntry[0];
          const rulesAreEquivalent = JSON.stringify(existingRule) === JSON.stringify(keyedRule);
          
          if (!rulesAreEquivalent) {
            throw new Error(`Key collision: The key '${keyedRule.key}' is used for different rule objects. One in 'repositoryItems' (or processed earlier) and another encountered in a 'patterns' array.`);
          }
          // If they are structurally equivalent, we can safely ignore this - it's the same rule definition
      }
      // If it's the same rule object or equivalent rule, we don't need to do anything - it's already in the repository
    }
    processedPatternsArray.push({ include: `#${keyedRule.key}` });
  }
  return processedPatternsArray;
}

// processRepository is no longer needed as finalRepositoryObject is built directly in processGrammar

function validateRegexp(regexp: string | undefined, node: any, prop: string, options: EmitOptions) {
  if (regexp === undefined) return;
  try {
    new OnigRegExp(regexp).testSync("");
  } catch (err: any) {
    let processedError = err;
    if (/^[0-9,]+$/.test(err.message)) {
      const array = new Uint8Array(err.message.split(",").map((s: string) => Number(s)));
      const buffer = Buffer.from(array);
      processedError = new Error(buffer.toString("utf-8"));
    }
    const sourceFile = options.errorSourceFilePath ?? "unknown_file";
    // eslint-disable-next-line no-console
    console.error(`${sourceFile}(1,1): error TM0001: Bad regex: ${JSON.stringify({[prop]: regexp})}: ${processedError.message}`);
    // eslint-disable-next-line no-console
    console.error(node);
    throw processedError;
  }
} 