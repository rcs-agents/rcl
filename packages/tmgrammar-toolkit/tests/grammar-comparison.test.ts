/**
 * Grammar Comparison Tests
 * 
 * Tests that compare grammars generated by tmgrammar-toolkit with those 
 * generated by the original tmlanguage-generator package.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import { readFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { grammar as bicepGrammar } from './fixtures/grammar-comparison/bicep.ts';
import { grammar as typespecGrammar } from './fixtures/grammar-comparison/typespec.ts';
import { emitJSON } from '../src/emit.js';

describe('Grammar Comparison Tests', () => {
  const fixturesPath = resolve(__dirname, 'fixtures/grammar-comparison');
  
  let originalBicepGrammar: any;
  let originalTypespecGrammar: any;
  let textmateToolkitBicepGrammar: any;
  let textmateToolkitTypespecGrammar: any;

  beforeAll(async () => {
    // Load original tmlanguage JSON files
    const bicepJsonPath = resolve(fixturesPath, 'bicep.tmlanguage.json');
    const typespecJsonPath = resolve(fixturesPath, 'typescpec.tmlanguage.json'); // Note: original has typo
    
    const bicepJson = await readFile(bicepJsonPath, 'utf-8');
    const typespecJson = await readFile(typespecJsonPath, 'utf-8');
    
    originalBicepGrammar = JSON.parse(bicepJson);
    originalTypespecGrammar = JSON.parse(typespecJson);
    
    // Generate tmgrammar-toolkit outputs using emitJSON
    const bicepGrammarJson = await emitJSON(bicepGrammar);
    const typespecGrammarJson = await emitJSON(typespecGrammar);
    
    textmateToolkitBicepGrammar = JSON.parse(bicepGrammarJson);
    textmateToolkitTypespecGrammar = JSON.parse(typespecGrammarJson);
  });

  describe('Bicep Grammar Comparison', () => {
    it('should have matching schema', () => {
      expect(textmateToolkitBicepGrammar.$schema).toBe(originalBicepGrammar.$schema);
    });

    it('should have matching name', () => {
      expect(textmateToolkitBicepGrammar.name).toBe(originalBicepGrammar.name);
    });

    it('should have matching scopeName', () => {
      expect(textmateToolkitBicepGrammar.scopeName).toBe(originalBicepGrammar.scopeName);
    });

    it('should have matching fileTypes', () => {
      expect(textmateToolkitBicepGrammar.fileTypes).toEqual(originalBicepGrammar.fileTypes);
    });

    it('should have a repository with rules', () => {
      expect(textmateToolkitBicepGrammar.repository).toBeDefined();
      expect(typeof textmateToolkitBicepGrammar.repository).toBe('object');
      expect(Object.keys(textmateToolkitBicepGrammar.repository).length).toBeGreaterThan(0);
    });

    it('should have patterns that reference repository rules', () => {
      expect(textmateToolkitBicepGrammar.patterns).toBeDefined();
      expect(Array.isArray(textmateToolkitBicepGrammar.patterns)).toBe(true);
      expect(textmateToolkitBicepGrammar.patterns.length).toBeGreaterThan(0);
    });

    it('should generate valid TextMate grammar structure', () => {
      // Check basic structure
      expect(textmateToolkitBicepGrammar).toHaveProperty('$schema');
      expect(textmateToolkitBicepGrammar).toHaveProperty('name');
      expect(textmateToolkitBicepGrammar).toHaveProperty('scopeName');
      expect(textmateToolkitBicepGrammar).toHaveProperty('fileTypes');
      expect(textmateToolkitBicepGrammar).toHaveProperty('patterns');
      expect(textmateToolkitBicepGrammar).toHaveProperty('repository');
    });

    it('should have repository rules with correct structure', () => {
      const repository = textmateToolkitBicepGrammar.repository;
      
      // Check that each rule has proper structure
      for (const [key, rule] of Object.entries(repository)) {
        expect(typeof key).toBe('string');
        expect(rule).toBeDefined();
        
        // Each rule should have either match, begin/end, or patterns
        const hasMatch = 'match' in (rule as object);
        const hasBeginEnd = 'begin' in (rule as object) && 'end' in (rule as object);
        const hasPatterns = 'patterns' in (rule as object);
        
        expect(hasMatch || hasBeginEnd || hasPatterns).toBe(true);
      }
    });

    it('should have similar repository structure to original', () => {
      const originalKeys = Object.keys(originalBicepGrammar.repository).sort();
      const toolkitKeys = Object.keys(textmateToolkitBicepGrammar.repository).sort();
      
      // We should have at least the same number of rules as the original
      expect(toolkitKeys.length).toBeGreaterThanOrEqual(originalKeys.length * 0.8); // Allow some variance
    });
  });

  describe('TypeSpec Grammar Comparison', () => {
    it('should have matching schema', () => {
      expect(textmateToolkitTypespecGrammar.$schema).toBe(originalTypespecGrammar.$schema);
    });

    it('should have matching name', () => {
      expect(textmateToolkitTypespecGrammar.name).toBe(originalTypespecGrammar.name);
    });

    it('should have matching scopeName', () => {
      expect(textmateToolkitTypespecGrammar.scopeName).toBe(originalTypespecGrammar.scopeName);
    });

    it('should have matching fileTypes', () => {
      expect(textmateToolkitTypespecGrammar.fileTypes).toEqual(originalTypespecGrammar.fileTypes);  
    });

    it('should have a repository with rules', () => {
      expect(textmateToolkitTypespecGrammar.repository).toBeDefined();
      expect(typeof textmateToolkitTypespecGrammar.repository).toBe('object');
      expect(Object.keys(textmateToolkitTypespecGrammar.repository).length).toBeGreaterThan(0);
    });

    it('should have patterns that reference repository rules', () => {
      expect(textmateToolkitTypespecGrammar.patterns).toBeDefined();
      expect(Array.isArray(textmateToolkitTypespecGrammar.patterns)).toBe(true);
      expect(textmateToolkitTypespecGrammar.patterns.length).toBeGreaterThan(0);
    });

    it('should generate valid TextMate grammar structure', () => {
      // Check basic structure
      expect(textmateToolkitTypespecGrammar).toHaveProperty('$schema');
      expect(textmateToolkitTypespecGrammar).toHaveProperty('name');
      expect(textmateToolkitTypespecGrammar).toHaveProperty('scopeName');
      expect(textmateToolkitTypespecGrammar).toHaveProperty('fileTypes');
      expect(textmateToolkitTypespecGrammar).toHaveProperty('patterns');
      expect(textmateToolkitTypespecGrammar).toHaveProperty('repository');
    });

    it('should have repository rules with correct structure', () => {
      const repository = textmateToolkitTypespecGrammar.repository;
      
      // Check that each rule has proper structure
      for (const [key, rule] of Object.entries(repository)) {
        expect(typeof key).toBe('string');
        expect(rule).toBeDefined();
        
        // Each rule should have either match, begin/end, or patterns
        const hasMatch = 'match' in (rule as object);
        const hasBeginEnd = 'begin' in (rule as object) && 'end' in (rule as object);
        const hasPatterns = 'patterns' in (rule as object);
        
        expect(hasMatch || hasBeginEnd || hasPatterns).toBe(true);
      }
    });

    it('should have similar repository structure to original', () => {
      const originalKeys = Object.keys(originalTypespecGrammar.repository).sort();
      const toolkitKeys = Object.keys(textmateToolkitTypespecGrammar.repository).sort();
      
      // We should have at least some of the same rules as the original
      expect(toolkitKeys.length).toBeGreaterThanOrEqual(originalKeys.length * 0.3); // TypeSpec conversion is partial
    });
  });

  describe('Cross-Grammar Structure Validation', () => {
    it('should generate grammars with consistent structure', () => {
      const bicepKeys = Object.keys(textmateToolkitBicepGrammar).sort();
      const typespecKeys = Object.keys(textmateToolkitTypespecGrammar).sort();
      
      // Both should have the same top-level structure
      expect(bicepKeys).toEqual(typespecKeys);
    });

    it('should have non-empty repositories', () => {
      expect(Object.keys(textmateToolkitBicepGrammar.repository).length).toBeGreaterThan(0);
      expect(Object.keys(textmateToolkitTypespecGrammar.repository).length).toBeGreaterThan(0);
    });

    it('should have all repository rules accessible', () => {
      const bicepRepo = textmateToolkitBicepGrammar.repository;
      const typespecRepo = textmateToolkitTypespecGrammar.repository;
      
      // Each grammar should have at least some rules in repository
      expect(Object.keys(bicepRepo).length).toBeGreaterThan(5);
      expect(Object.keys(typespecRepo).length).toBeGreaterThan(5);
    });
  });

  describe('JSON Output Comparison', () => {
    it('should generate valid JSON that matches original structure', async () => {
      // Test that our generated JSON is parseable and structurally similar
      const bicepJson = await emitJSON(bicepGrammar);
      const typespecJson = await emitJSON(typespecGrammar);
      
      expect(() => JSON.parse(bicepJson)).not.toThrow();
      expect(() => JSON.parse(typespecJson)).not.toThrow();
      
      const parsedBicep = JSON.parse(bicepJson);
      const parsedTypespec = JSON.parse(typespecJson);
      
      // Check that the parsed JSON has all required fields
      expect(parsedBicep).toHaveProperty('$schema');
      expect(parsedBicep).toHaveProperty('name');
      expect(parsedBicep).toHaveProperty('scopeName');
      expect(parsedBicep).toHaveProperty('fileTypes');
      expect(parsedBicep).toHaveProperty('patterns');
      expect(parsedBicep).toHaveProperty('repository');
      
      expect(parsedTypespec).toHaveProperty('$schema');
      expect(parsedTypespec).toHaveProperty('name');
      expect(parsedTypespec).toHaveProperty('scopeName');
      expect(parsedTypespec).toHaveProperty('fileTypes');
      expect(parsedTypespec).toHaveProperty('patterns');
      expect(parsedTypespec).toHaveProperty('repository');
    });
  });

  describe('Regression Tests', () => {
    it('should maintain consistent grammar structure over time', () => {
      // Save current structure for regression testing
      const bicepStructure = {
        schema: textmateToolkitBicepGrammar.$schema,
        name: textmateToolkitBicepGrammar.name,
        scopeName: textmateToolkitBicepGrammar.scopeName,
        fileTypes: textmateToolkitBicepGrammar.fileTypes,
        repositoryKeys: Object.keys(textmateToolkitBicepGrammar.repository).sort(),
        patternsLength: textmateToolkitBicepGrammar.patterns.length
      };

      const typespecStructure = {
        schema: textmateToolkitTypespecGrammar.$schema,
        name: textmateToolkitTypespecGrammar.name,
        scopeName: textmateToolkitTypespecGrammar.scopeName,
        fileTypes: textmateToolkitTypespecGrammar.fileTypes,
        repositoryKeys: Object.keys(textmateToolkitTypespecGrammar.repository).sort(),
        patternsLength: textmateToolkitTypespecGrammar.patterns.length
      };

      // These snapshots will catch unintentional changes
      expect(bicepStructure).toMatchSnapshot('bicep-grammar-structure');
      expect(typespecStructure).toMatchSnapshot('typespec-grammar-structure');
    });
  });
}); 