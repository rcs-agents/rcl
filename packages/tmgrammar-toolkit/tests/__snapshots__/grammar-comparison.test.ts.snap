// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Grammar Comparison Tests Regression Tests should maintain consistent grammar structure over time: bicep-grammar-structure 1`] = `
{
  "fileTypes": [
    ".bicep",
    ".bicepparam",
  ],
  "name": "Bicep",
  "patternsLength": 2,
  "repositoryKeys": [
    "array-literal",
    "block-comment",
    "comments",
    "decorator",
    "directive",
    "directive-variable",
    "escape-character",
    "expression",
    "function-call",
    "identifier",
    "keyword",
    "lambda-start",
    "line-comment",
    "named-literal",
    "numeric-literal",
    "object-literal",
    "object-property-key",
    "string-literal",
    "string-literal-subst",
    "string-verbatim",
  ],
  "schema": "https://raw.githubusercontent.com/martinring/tmlanguage/master/tmlanguage.json",
  "scopeName": "source.bicep",
}
`;

exports[`Grammar Comparison Tests Regression Tests should maintain consistent grammar structure over time: typespec-grammar-structure 1`] = `
{
  "fileTypes": [
    "tsp",
  ],
  "name": "TypeSpec",
  "patternsLength": 1,
  "repositoryKeys": [
    "block-comment",
    "boolean-literal",
    "callExpression",
    "directive",
    "doc-comment",
    "doc-comment-block",
    "doc-comment-param",
    "doc-comment-return-tag",
    "doc-comment-unknown-tag",
    "escape-character",
    "expression",
    "identifier-expression",
    "line-comment",
    "numeric-literal",
    "operator-assignment",
    "parenthesized-expression",
    "punctuation-accessor",
    "punctuation-comma",
    "punctuation-semicolon",
    "statement",
    "string-literal",
    "template-expression",
    "token",
    "triple-quoted-string-literal",
    "type-parameter",
    "type-parameter-constraint",
    "type-parameter-default",
    "type-parameters",
  ],
  "schema": "https://raw.githubusercontent.com/martinring/tmlanguage/master/tmlanguage.json",
  "scopeName": "source.tsp",
}
`;
